core/src/main/scala/org/apache/spark/rdd/ParallelCollectionRDD.scala

```diff
@@ -127,18 +127,15 @@ private object ParallelCollectionRDD {
       })
     }
     seq match {
-      case r: Range.Inclusive => {
-        val sign = if (r.step < 0) {
-          -1
-        } else {
-          1
-        }
-        slice(new Range(
-          r.start, r.end + sign, r.step).asInstanceOf[Seq[T]], numSlices)
-      }
       case r: Range => {
-        positions(r.length, numSlices).map({
-          case (start, end) =>
+        // 1 to Int.MaxValue and (-2 to Int.MinValue by -1) can trigger exclusive range int overflow
+        val needsInclusiveRange = r.isInclusive && (r.end == Int.MaxValue || r.end == Int.MinValue)
+        positions(r.length, numSlices).zipWithIndex.map({
+          // we need an inclusive range to avoid int overflow and setting the last range to be
+          // inclusive is suffice 
+          case ((start, end), index) if needsInclusiveRange && index == numSlices - 1 =>
+            new Range.Inclusive(r.start + start * r.step, r.end, r.step)
+          case ((start, end), _) =>
             new Range(r.start + start * r.step, r.start + end * r.step, r.step)
```

Comment:
No doubt that your version is more straightforward than mine. When I wrote my code, I didn't consider splitting normal inclusive range using inclusive range. However the benefit of my implementation is that the splitting result will be same as in the master for normal inclusive ranges. I wonder there may be some spark code rely on the exclusive range output.  And of course, I think we should update the corresponding document for this kind of change.

I will covert the pattern matching to one case and update the implementation when we decided which one fits better.